text/0225-configuration.md

```diff
@@ -0,0 +1,330 @@
+
+### ParseAndValidateConfigurationFromFile(filepath, format) -> config
+
+An API called `ParseAndValidateConfigurationFromFile` receives a string parameter indicating the file path containing the configuration to be parsed. An optional format parameter may be provided to indicate the format that this configuration uses. YAML *SHOULD* be supported, but JSON, which is a strict subset of YAML, *MAY* be supported if a language ecosystem has good reason. For example a language that has built-in support for JSON running in an environment that is sensitive to additional dependencies *MAY* only be able to support JSON configuration.
+
+The method returns a `Configuration` model that has been validated. This API *MAY* return an error or raise an exception, whichever is idiomatic to the implementation for the following reasons:
+
+* file doesn't exist or is invalid
+* configuration parsed is invalid
+
+#### Python ParseAndValidateConfigurationFromFile example
+
+```python
+
+filepath = "./config.yaml"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filepath)
+except Exception as e:
+  print(e)
+
+filename = "./config.json"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filename, format="json")
+except Exception as e:
+  raise e
+
+```
+
+#### Go ParseAndValidateConfigurationFromFile example
+
+```go
+
+filename := "./config.yaml"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename)
+if err != nil {
+  return err
+}
+
+filename := "./config.json"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename, otelconfig.WithFormat("json"))
+if err != nil {
+  return err
+}
+
+```
+
+Implementations *MUST* allow users to specify an environment variable to set the configuration file. This gives flexibility to end users of implementations that do not support command line arguments. The proposed name for this variable:
+
+* `OTEL_CONFIG_FILE`
+
+### Configuration model
+
+To allow SDKs and instrumentation libraries to accept configuration without having to implement the parsing logic, a `Configuration` model *MAY* be provided by implementations. This object:
+
+* has already been parsed from a file or data structure
+* is structurally valid (errors may yet occur when SDK or instrumentation interprets the object)
+
+### Configuration file
+
+The following demonstrates an example of a configuration file format (full example [here](https://github.com/MrAlias/otel-schema/blob/main/config.yaml)):
+
+```yaml
+# include version specification in configuration files to help with parsing and schema evolution.
+file_format: 0.1
+sdk:
```

Comment:
Just noticed this property. Does it imply that this file will be shared with other non-sdk components?

Reply:
A goal of this is to hold configuration for SDKs as well as instrumentation. However, there are some challenges with configuring instrumentation:

- Instrumentation configuration is non standard across libraries and across languages. How much common configuration is there? The answer may be little, and that we support instrumentation configuration in whatever way we support extension components (custom exporters, processors, samplers, etc). 
- How is configuration delivered to instrumentation? Some languages have centralized setup of instrumentation, but would need to accommodate situations where "library" or "native" instrumentation configured manually and sprinkled throughout the application. Maybe there's a configuration model we pass around?
- We can't guarantee that instrumentation will adhere to the configuration. For example, if we define standard options to enable/disable instrumentation libraries, users will come to expect that to work. However, we won't have any way to control native instrumentation or instrumentation written by third parties. Giving the impression that centralized configuration of instrumentation through a file is possible may cause more problems than it solves.

Reply:
so the `sdk` key is used in order to allow space for additional keys in the future for instrumentation, but those are out of scope for now

Reply:
Correct.

It could end up being the case that we want to have the SDK keys (`tracer_provider`, `meter_provider`, etc) at the top level, and add an additional top level key for `instrumentation` config when appropriate. 

But we can hash out the exact details of the file config schema later when added to the spec. This is just an example of file configuration for the purposes of this OTEP. @MrAlias is https://github.com/open-telemetry/oteps/pull/225#discussion_r1142339164 text saying that the exact schema of the config file is an open question.

Reply:
I've tried to capture the issues in this thread in the open questions section to be addressed in the specification (c7e738b109feb6643c3ba274f7d0506efdf10b50). Therefore, I'm going to resolve this as it being tracked.

Please take a look a the changes, and reopen the discussion if you think this was in error.